// astronomia/es/saturnmoons.js
/**
 * @copyright 2013 Sonia Keys
 * @copyright 2016 commenthol
 * @license MIT
 * @module saturnmoons
 */

/**
 * Saturnmoons: Chapter 46, Positions of the Satellites of Saturn
 */
import base from './base';
import coord from './coord';
import planetposition from './planetposition';
import precess from './precess';
import solar from './solar'; // array positions of Saturnmoons returned from positions().

export var mimas = 0;
export var enceladus = 1;
export var tethys = 2;
export var dione = 3;
export var rhea = 4;
export var titan = 5;
export var hyperion = 6;
export var iapetus = 7;
/**
 * XY holds coordinates returned from positions().
 */

function XY(x, y) {
  this.x = x;
  this.y = y;
}

var d = Math.PI / 180;
/**
 * Positions returns positions of the eight major moons of Saturn.
 *
 * Results returned in argument pos, which must not be undefined.
 *
 * Result units are Saturn radii.
 *
 * @param {number} jde - Julian ephemeris day
 * @param {planetposition.Planet} earth - VSOP87 planet Earth
 * @param {planetposition.Planet} saturn - VSOP87 planet Saturn
 * @return {XY[]} Array of Moon Positions in `XY`
 *   Use `M.mimas ... M.iapetus` to resolve to Moon and its position at `jde`
 */

export function positions(jde, earth, saturn) {
  var sol = solar.trueVSOP87(earth, jde);
  var _ref = [sol.lon, sol.lat, sol.range],
      s = _ref[0],
      β = _ref[1],
      R = _ref[2];

  var _base$sincos = base.sincos(s),
      ss = _base$sincos[0],
      cs = _base$sincos[1];

  var sβ = Math.sin(β);
  var Δ = 9.0;

  var x, y, z, _jde;

  var f = function f() {
    var τ = base.lightTime(Δ);
    _jde = jde - τ;

    var _saturn$position = saturn.position(_jde),
        lon = _saturn$position.lon,
        lat = _saturn$position.lat,
        range = _saturn$position.range;

    var fk5 = planetposition.toFK5(lon, lat, _jde);
    var _ref2 = [fk5.lon, fk5.lat],
        l = _ref2[0],
        b = _ref2[1];

    var _base$sincos2 = base.sincos(l),
        sl = _base$sincos2[0],
        cl = _base$sincos2[1];

    var _base$sincos3 = base.sincos(b),
        sb = _base$sincos3[0],
        cb = _base$sincos3[1];

    x = range * cb * cl + R * cs;
    y = range * cb * sl + R * ss;
    z = range * sb + R * sβ;
    Δ = Math.sqrt(x * x + y * y + z * z);
  };

  f();
  f();
  var λ0 = Math.atan2(y, x);
  var β0 = Math.atan(z / Math.hypot(x, y));
  var ecl = new coord.Ecliptic(λ0, β0);
  ecl = precess.eclipticPosition(ecl, base.JDEToJulianYear(jde), base.JDEToJulianYear(base.B1950), 0, 0);
  λ0 = ecl.lon;
  β0 = ecl.lat;
  var q = new Qs(_jde);
  var s4 = [new R4(), // 0 unused
  q.mimas(), q.enceladus(), q.tethys(), q.dione(), q.rhea(), q.titan(), q.hyperion(), q.iapetus()];
  var X = new Array(9).fill(0);
  var Y = new Array(9).fill(0);
  var Z = new Array(9).fill(0);

  for (var j = 1; j <= 8; j++) {
    var u = s4[j].λ - s4[j].Ω;
    var w = s4[j].Ω - 168.8112 * d;

    var _base$sincos4 = base.sincos(u),
        su = _base$sincos4[0],
        cu = _base$sincos4[1];

    var _base$sincos5 = base.sincos(w),
        sw = _base$sincos5[0],
        cw = _base$sincos5[1];

    var _base$sincos6 = base.sincos(s4[j].γ),
        sγ = _base$sincos6[0],
        cγ = _base$sincos6[1];

    var r = s4[j].r;
    X[j] = r * (cu * cw - su * cγ * sw);
    Y[j] = r * (su * cw * cγ + cu * sw);
    Z[j] = r * su * sγ;
  }

  Z[0] = 1;

  var _base$sincos7 = base.sincos(λ0),
      sλ0 = _base$sincos7[0],
      cλ0 = _base$sincos7[1];

  var _base$sincos8 = base.sincos(β0),
      sβ0 = _base$sincos8[0],
      cβ0 = _base$sincos8[1];

  var A = new Array(9).fill(0);
  var B = new Array(9).fill(0);
  var C = new Array(9).fill(0);

  for (var _j in X) {
    var a = X[_j];
    var b = q.c1 * Y[_j] - q.s1 * Z[_j];
    var c = q.s1 * Y[_j] + q.c1 * Z[_j];
    var a0 = q.c2 * a - q.s2 * b;
    b = q.s2 * a + q.c2 * b;
    a = a0;
    A[_j] = a * sλ0 - b * cλ0;
    b = a * cλ0 + b * sλ0;
    B[_j] = b * cβ0 + c * sβ0;
    C[_j] = c * cβ0 - b * sβ0;
  }

  var pos = new Array(9);
  var D = Math.atan2(A[0], C[0]);

  var _base$sincos9 = base.sincos(D),
      sD = _base$sincos9[0],
      cD = _base$sincos9[1];

  for (var _j2 = 1; _j2 <= 8; _j2++) {
    X[_j2] = A[_j2] * cD - C[_j2] * sD;
    Y[_j2] = A[_j2] * sD + C[_j2] * cD;
    Z[_j2] = B[_j2];

    var _d = X[_j2] / s4[_j2].r;

    X[_j2] += Math.abs(Z[_j2]) / k[_j2] * Math.sqrt(1 - _d * _d);
    var W = Δ / (Δ + Z[_j2] / 2475);
    pos[_j2 - 1] = new XY(X[_j2] * W, Y[_j2] * W);
  }

  return pos;
}
var k = [0, 20947, 23715, 26382, 29876, 35313, 53800, 59222, 91820];

function R4(λ, r, γ, Ω) {
  this.λ = λ || 0;
  this.r = r || 0;
  this.γ = γ || 0;
  this.Ω = Ω || 0;
}

export function Qs(jde) {
  this.t1 = jde - 2411093;
  this.t2 = this.t1 / 365.25;
  this.t3 = (jde - 2433282.423) / 365.25 + 1950;
  this.t4 = jde - 2411368;
  this.t5 = this.t4 / 365.25;
  this.t6 = jde - 2415020;
  this.t7 = this.t6 / 36525;
  this.t8 = this.t6 / 365.25;
  this.t9 = (jde - 2442000.5) / 365.25;
  this.t10 = jde - 2409786;
  this.t11 = this.t10 / 36525;
  this.W0 = 5.095 * d * (this.t3 - 1866.39);
  this.W1 = 74.4 * d + 32.39 * d * this.t2;
  this.W2 = 134.3 * d + 92.62 * d * this.t2;
  this.W3 = 42 * d - 0.5118 * d * this.t5;
  this.W4 = 276.59 * d + 0.5118 * d * this.t5;
  this.W5 = 267.2635 * d + 1222.1136 * d * this.t7;
  this.W6 = 175.4762 * d + 1221.5515 * d * this.t7;
  this.W7 = 2.4891 * d + 0.002435 * d * this.t7;
  this.W8 = 113.35 * d - 0.2597 * d * this.t7;
  this.s1 = Math.sin(28.0817 * d);
  this.c1 = Math.cos(28.0817 * d);
  this.s2 = Math.sin(168.8112 * d);
  this.c2 = Math.cos(168.8112 * d);
  this.e1 = 0.05589 - 0.000346 * this.t7;
  this.sW0 = Math.sin(this.W0);
  this.s3W0 = Math.sin(3 * this.W0);
  this.s5W0 = Math.sin(5 * this.W0);
  this.sW1 = Math.sin(this.W1);
  this.sW2 = Math.sin(this.W2);
  this.sW3 = Math.sin(this.W3);
  this.cW3 = Math.cos(this.W3);
  this.sW4 = Math.sin(this.W4);
  this.cW4 = Math.cos(this.W4);
  this.sW7 = Math.sin(this.W7);
  this.cW7 = Math.cos(this.W7);
  return this;
}

Qs.prototype.mimas = function () {
  var r = new R4();
  var L = 127.64 * d + 381.994497 * d * this.t1 - 43.57 * d * this.sW0 - 0.72 * d * this.s3W0 - 0.02144 * d * this.s5W0;
  var p = 106.1 * d + 365.549 * d * this.t2;
  var M = L - p;
  var C = 2.18287 * d * Math.sin(M) + 0.025988 * d * Math.sin(2 * M) + 0.00043 * d * Math.sin(3 * M);
  r.λ = L + C;
  r.r = 3.06879 / (1 + 0.01905 * Math.cos(M + C));
  r.γ = 1.563 * d;
  r.Ω = 54.5 * d - 365.072 * d * this.t2;
  return r;
};

Qs.prototype.enceladus = function () {
  var r = new R4();
  var L = 200.317 * d + 262.7319002 * d * this.t1 + 0.25667 * d * this.sW1 + 0.20883 * d * this.sW2;
  var p = 309.107 * d + 123.44121 * d * this.t2;
  var M = L - p;
  var C = 0.55577 * d * Math.sin(M) + 0.00168 * d * Math.sin(2 * M);
  r.λ = L + C;
  r.r = 3.94118 / (1 + 0.00485 * Math.cos(M + C));
  r.γ = 0.0262 * d;
  r.Ω = 348 * d - 151.95 * d * this.t2;
  return r;
};

Qs.prototype.tethys = function () {
  var r = new R4();
  r.λ = 285.306 * d + 190.69791226 * d * this.t1 + 2.063 * d * this.sW0 + 0.03409 * d * this.s3W0 + 0.001015 * d * this.s5W0;
  r.r = 4.880998;
  r.γ = 1.0976 * d;
  r.Ω = 111.33 * d - 72.2441 * d * this.t2;
  return r;
};

Qs.prototype.dione = function () {
  var r = new R4();
  var L = 254.712 * d + 131.53493193 * d * this.t1 - 0.0215 * d * this.sW1 - 0.01733 * d * this.sW2;
  var p = 174.8 * d + 30.82 * d * this.t2;
  var M = L - p;
  var C = 0.24717 * d * Math.sin(M) + 0.00033 * d * Math.sin(2 * M);
  r.λ = L + C;
  r.r = 6.24871 / (1 + 0.002157 * Math.cos(M + C));
  r.γ = 0.0139 * d;
  r.Ω = 232 * d - 30.27 * d * this.t2;
  return r;
};

Qs.prototype.rhea = function () {
  var pʹ = 342.7 * d + 10.057 * d * this.t2;

  var _base$sincos10 = base.sincos(pʹ),
      spʹ = _base$sincos10[0],
      cpʹ = _base$sincos10[1];

  var a1 = 0.000265 * spʹ + 0.001 * this.sW4;
  var a2 = 0.000265 * cpʹ + 0.001 * this.cW4;
  var e = Math.hypot(a1, a2);
  var p = Math.atan2(a1, a2);
  var N = 345 * d - 10.057 * d * this.t2;

  var _base$sincos11 = base.sincos(N),
      sN = _base$sincos11[0],
      cN = _base$sincos11[1];

  var λʹ = 359.244 * d + 79.6900472 * d * this.t1 + 0.086754 * d * sN;
  var i = 28.0362 * d + 0.346898 * d * cN + 0.0193 * d * this.cW3;
  var Ω = 168.8034 * d + 0.736936 * d * sN + 0.041 * d * this.sW3;
  var a = 8.725924;
  return this.subr(λʹ, p, e, a, Ω, i);
};

Qs.prototype.subr = function (λʹ, p, e, a, Ω, i) {
  var r = new R4();
  var M = λʹ - p;
  var e2 = e * e;
  var e3 = e2 * e;
  var e4 = e2 * e2;
  var e5 = e3 * e2;
  var C = (2 * e - 0.25 * e3 + 0.0520833333 * e5) * Math.sin(M) + (1.25 * e2 - 0.458333333 * e4) * Math.sin(2 * M) + (1.083333333 * e3 - 0.671875 * e5) * Math.sin(3 * M) + 1.072917 * e4 * Math.sin(4 * M) + 1.142708 * e5 * Math.sin(5 * M);
  r.r = a * (1 - e2) / (1 + e * Math.cos(M + C)); // return value

  var g = Ω - 168.8112 * d;

  var _base$sincos12 = base.sincos(i),
      si = _base$sincos12[0],
      ci = _base$sincos12[1];

  var _base$sincos13 = base.sincos(g),
      sg = _base$sincos13[0],
      cg = _base$sincos13[1];

  var a1 = si * sg;
  var a2 = this.c1 * si * cg - this.s1 * ci;
  r.γ = Math.asin(Math.hypot(a1, a2)); // return value

  var u = Math.atan2(a1, a2);
  r.Ω = 168.8112 * d + u; // return value (w)

  var h = this.c1 * si - this.s1 * ci * cg;
  var ψ = Math.atan2(this.s1 * sg, h);
  r.λ = λʹ + C + u - g - ψ; // return value

  return r;
};

Qs.prototype.titan = function () {
  var _this = this;

  var L = 261.1582 * d + 22.57697855 * d * this.t4 + 0.074025 * d * this.sW3;
  var iʹ = 27.45141 * d + 0.295999 * d * this.cW3;
  var Ωʹ = 168.66925 * d + 0.628808 * d * this.sW3;

  var _base$sincos14 = base.sincos(iʹ),
      siʹ = _base$sincos14[0],
      ciʹ = _base$sincos14[1];

  var _base$sincos15 = base.sincos(Ωʹ - this.W8),
      sΩʹW8 = _base$sincos15[0],
      cΩʹW8 = _base$sincos15[1];

  var a1 = this.sW7 * sΩʹW8;
  var a2 = this.cW7 * siʹ - this.sW7 * ciʹ * cΩʹW8;
  var g0 = 102.8623 * d;
  var ψ = Math.atan2(a1, a2);
  var s = Math.hypot(a1, a2);
  var g = this.W4 - Ωʹ - ψ;
  var ϖ = 0;

  var _base$sincos16 = base.sincos(2 * g0),
      s2g0 = _base$sincos16[0],
      c2g0 = _base$sincos16[1];

  var f = function f() {
    ϖ = _this.W4 + 0.37515 * d * (Math.sin(2 * g) - s2g0);
    g = ϖ - Ωʹ - ψ;
  };

  f();
  f();
  f();
  var eʹ = 0.029092 + 0.00019048 * (Math.cos(2 * g) - c2g0);
  var qq = 2 * (this.W5 - ϖ);
  var b1 = siʹ * sΩʹW8;
  var b2 = this.cW7 * siʹ * cΩʹW8 - this.sW7 * ciʹ;
  var θ = Math.atan2(b1, b2) + this.W8;

  var _base$sincos17 = base.sincos(qq),
      sq = _base$sincos17[0],
      cq = _base$sincos17[1];

  var e = eʹ + 0.002778797 * eʹ * cq;
  var p = ϖ + 0.159215 * d * sq;
  var u = 2 * this.W5 - 2 * θ + ψ;

  var _base$sincos18 = base.sincos(u),
      su = _base$sincos18[0],
      cu = _base$sincos18[1];

  var h = 0.9375 * eʹ * eʹ * sq + 0.1875 * s * s * Math.sin(2 * (this.W5 - θ));
  var λʹ = L - 0.254744 * d * (this.e1 * Math.sin(this.W6) + 0.75 * this.e1 * this.e1 * Math.sin(2 * this.W6) + h);
  var i = iʹ + 0.031843 * d * s * cu;
  var Ω = Ωʹ + 0.031843 * d * s * su / siʹ;
  var a = 20.216193;
  return this.subr(λʹ, p, e, a, Ω, i);
};

Qs.prototype.hyperion = function () {
  var η = 92.39 * d + 0.5621071 * d * this.t6;
  var ζ = 148.19 * d - 19.18 * d * this.t8;
  var θ = 184.8 * d - 35.41 * d * this.t9;
  var θʹ = θ - 7.5 * d;
  var as = 176 * d + 12.22 * d * this.t8;
  var bs = 8 * d + 24.44 * d * this.t8;
  var cs = bs + 5 * d;
  var ϖ = 69.898 * d - 18.67088 * d * this.t8;
  var φ = 2 * (ϖ - this.W5);
  var χ = 94.9 * d - 2.292 * d * this.t8;

  var _base$sincos19 = base.sincos(η),
      sη = _base$sincos19[0],
      cη = _base$sincos19[1];

  var _base$sincos20 = base.sincos(ζ),
      sζ = _base$sincos20[0],
      cζ = _base$sincos20[1];

  var _base$sincos21 = base.sincos(2 * ζ),
      s2ζ = _base$sincos21[0],
      c2ζ = _base$sincos21[1];

  var _base$sincos22 = base.sincos(3 * ζ),
      s3ζ = _base$sincos22[0],
      c3ζ = _base$sincos22[1];

  var _base$sincos23 = base.sincos(ζ + η),
      sζpη = _base$sincos23[0],
      cζpη = _base$sincos23[1];

  var _base$sincos24 = base.sincos(ζ - η),
      sζmη = _base$sincos24[0],
      cζmη = _base$sincos24[1];

  var _base$sincos25 = base.sincos(φ),
      sφ = _base$sincos25[0],
      cφ = _base$sincos25[1];

  var _base$sincos26 = base.sincos(χ),
      sχ = _base$sincos26[0],
      cχ = _base$sincos26[1];

  var _base$sincos27 = base.sincos(cs),
      scs = _base$sincos27[0],
      ccs = _base$sincos27[1];

  var a = 24.50601 - 0.08686 * cη - 0.00166 * cζpη + 0.00175 * cζmη;
  var e = 0.103458 - 0.004099 * cη - 0.000167 * cζpη + 0.000235 * cζmη + 0.02303 * cζ - 0.00212 * c2ζ + 0.000151 * c3ζ + 0.00013 * cφ;
  var p = ϖ + 0.15648 * d * sχ - 0.4457 * d * sη - 0.2657 * d * sζpη - 0.3573 * d * sζmη - 12.872 * d * sζ + 1.668 * d * s2ζ - 0.2419 * d * s3ζ - 0.07 * d * sφ;
  var λʹ = 177.047 * d + 16.91993829 * d * this.t6 + 0.15648 * d * sχ + 9.142 * d * sη + 0.007 * d * Math.sin(2 * η) - 0.014 * d * Math.sin(3 * η) + 0.2275 * d * sζpη + 0.2112 * d * sζmη - 0.26 * d * sζ - 0.0098 * d * s2ζ - 0.013 * d * Math.sin(as) + 0.017 * d * Math.sin(bs) - 0.0303 * d * sφ;
  var i = 27.3347 * d + 0.6434886 * d * cχ + 0.315 * d * this.cW3 + 0.018 * d * Math.cos(θ) - 0.018 * d * ccs;
  var Ω = 168.6812 * d + 1.40136 * d * cχ + 0.68599 * d * this.sW3 - 0.0392 * d * scs + 0.0366 * d * Math.sin(θʹ);
  return this.subr(λʹ, p, e, a, Ω, i);
};

Qs.prototype.iapetus = function () {
  var L = 261.1582 * d + 22.57697855 * d * this.t4;
  var ϖʹ = 91.796 * d + 0.562 * d * this.t7;
  var ψ = 4.367 * d - 0.195 * d * this.t7;
  var θ = 146.819 * d - 3.198 * d * this.t7;
  var φ = 60.47 * d + 1.521 * d * this.t7;
  var Φ = 205.055 * d - 2.091 * d * this.t7;
  var eʹ = 0.028298 + 0.001156 * this.t11;
  var ϖ0 = 352.91 * d + 11.71 * d * this.t11;
  var μ = 76.3852 * d + 4.53795125 * d * this.t10;
  var iʹ = base.horner(this.t11, 18.4602 * d, -0.9518 * d, -0.072 * d, 0.0054 * d);
  var Ωʹ = base.horner(this.t11, 143.198 * d, -3.919 * d, 0.116 * d, 0.008 * d);
  var l = μ - ϖ0;
  var g = ϖ0 - Ωʹ - ψ;
  var g1 = ϖ0 - Ωʹ - φ;
  var ls = this.W5 - ϖʹ;
  var gs = ϖʹ - θ;
  var lT = L - this.W4;
  var gT = this.W4 - Φ;
  var u1 = 2 * (l + g - ls - gs);
  var u2 = l + g1 - lT - gT;
  var u3 = l + 2 * (g - ls - gs);
  var u4 = lT + gT - g1;
  var u5 = 2 * (ls + gs);

  var _base$sincos28 = base.sincos(l),
      sl = _base$sincos28[0],
      cl = _base$sincos28[1];

  var _base$sincos29 = base.sincos(u1),
      su1 = _base$sincos29[0],
      cu1 = _base$sincos29[1];

  var _base$sincos30 = base.sincos(u2),
      su2 = _base$sincos30[0],
      cu2 = _base$sincos30[1];

  var _base$sincos31 = base.sincos(u3),
      su3 = _base$sincos31[0],
      cu3 = _base$sincos31[1];

  var _base$sincos32 = base.sincos(u4),
      su4 = _base$sincos32[0],
      cu4 = _base$sincos32[1];

  var _base$sincos33 = base.sincos(l + u2),
      slu2 = _base$sincos33[0],
      clu2 = _base$sincos33[1];

  var _base$sincos34 = base.sincos(g1 - gT),
      sg1gT = _base$sincos34[0],
      cg1gT = _base$sincos34[1];

  var _base$sincos35 = base.sincos(u5 - 2 * g),
      su52g = _base$sincos35[0],
      cu52g = _base$sincos35[1];

  var _base$sincos36 = base.sincos(u5 + ψ),
      su5ψ = _base$sincos36[0],
      cu5ψ = _base$sincos36[1];

  var _base$sincos37 = base.sincos(u2 + φ),
      su2φ = _base$sincos37[0],
      cu2φ = _base$sincos37[1];

  var _base$sincos38 = base.sincos(l + g1 + lT + gT + φ),
      s5 = _base$sincos38[0],
      c5 = _base$sincos38[1];

  var a = 58.935028 + 0.004638 * cu1 + 0.058222 * cu2;
  var e = eʹ - 0.0014097 * cg1gT + 0.0003733 * cu52g + 0.000118 * cu3 + 0.0002408 * cl + 0.0002849 * clu2 + 0.000619 * cu4;
  var w = 0.08077 * d * sg1gT + 0.02139 * d * su52g - 0.00676 * d * su3 + 0.0138 * d * sl + 0.01632 * d * slu2 + 0.03547 * d * su4;
  var p = ϖ0 + w / eʹ;
  var λʹ = μ - 0.04299 * d * su2 - 0.00789 * d * su1 - 0.06312 * d * Math.sin(ls) - 0.00295 * d * Math.sin(2 * ls) - 0.02231 * d * Math.sin(u5) + 0.0065 * d * su5ψ;
  var i = iʹ + 0.04204 * d * cu5ψ + 0.00235 * d * c5 + 0.0036 * d * cu2φ;
  var wʹ = 0.04204 * d * su5ψ + 0.00235 * d * s5 + 0.00358 * d * su2φ;
  var Ω = Ωʹ + wʹ / Math.sin(iʹ);
  return this.subr(λʹ, p, e, a, Ω, i);
};

export default {
  mimas: mimas,
  enceladus: enceladus,
  tethys: tethys,
  dione: dione,
  rhea: rhea,
  titan: titan,
  hyperion: hyperion,
  iapetus: iapetus,
  positions: positions,
  Qs: Qs
};
