// astronomia/lib/elliptic.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.position = position;
exports.astrometricJ2000 = astrometricJ2000;
exports.velocity = velocity;
exports.vAphelion = vAphelion;
exports.vPerihelion = vPerihelion;
exports.length1 = length1;
exports.length2 = length2;
exports.length4 = length4;
exports["default"] = exports.Elements = void 0;

var _apparent = _interopRequireDefault(require("./apparent"));

var _base = _interopRequireDefault(require("./base"));

var _coord = _interopRequireDefault(require("./coord"));

var _kepler = _interopRequireDefault(require("./kepler"));

var _nutation = _interopRequireDefault(require("./nutation"));

var _planetposition = _interopRequireDefault(require("./planetposition"));

var _solarxyz = _interopRequireDefault(require("./solarxyz"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/**
 * Position returns observed equatorial coordinates of a planet at a given time.
 *
 * Argument p must be a valid V87Planet object for the observed planet.
 * Argument earth must be a valid V87Planet object for Earth.
 *
 * Results are right ascension and declination, α and δ in radians.
 */
function position(planet, earth, jde) {
  // (p, earth *pp.V87Planet, jde float64)  (α, δ float64)
  var x;
  var y;
  var z;
  var posEarth = earth.position(jde);
  var _ref = [posEarth.lon, posEarth.lat, posEarth.range],
      L0 = _ref[0],
      B0 = _ref[1],
      R0 = _ref[2];

  var _base$sincos = _base["default"].sincos(B0),
      _base$sincos2 = _slicedToArray(_base$sincos, 2),
      sB0 = _base$sincos2[0],
      cB0 = _base$sincos2[1];

  var _base$sincos3 = _base["default"].sincos(L0),
      _base$sincos4 = _slicedToArray(_base$sincos3, 2),
      sL0 = _base$sincos4[0],
      cL0 = _base$sincos4[1];

  function pos() {
    var τ = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
    var pos = planet.position(jde - τ);
    var _ref2 = [pos.lon, pos.lat, pos.range],
        L = _ref2[0],
        B = _ref2[1],
        R = _ref2[2];

    var _base$sincos5 = _base["default"].sincos(B),
        _base$sincos6 = _slicedToArray(_base$sincos5, 2),
        sB = _base$sincos6[0],
        cB = _base$sincos6[1];

    var _base$sincos7 = _base["default"].sincos(L),
        _base$sincos8 = _slicedToArray(_base$sincos7, 2),
        sL = _base$sincos8[0],
        cL = _base$sincos8[1];

    x = R * cB * cL - R0 * cB0 * cL0;
    y = R * cB * sL - R0 * cB0 * sL0;
    z = R * sB - R0 * sB0;
  }

  pos();
  var Δ = Math.sqrt(x * x + y * y + z * z); // (33.4) p. 224

  var τ = _base["default"].lightTime(Δ); // repeating with jde-τ


  pos(τ);
  var λ = Math.atan2(y, x); // (33.1) p. 223

  var β = Math.atan2(z, Math.hypot(x, y)); // (33.2) p. 223

  var _apparent$eclipticAbe = _apparent["default"].eclipticAberration(λ, β, jde),
      _apparent$eclipticAbe2 = _slicedToArray(_apparent$eclipticAbe, 2),
      Δλ = _apparent$eclipticAbe2[0],
      Δβ = _apparent$eclipticAbe2[1];

  var fk5 = _planetposition["default"].toFK5(λ + Δλ, β + Δβ, jde);

  λ = fk5.lon;
  β = fk5.lat;

  var _nutation$nutation = _nutation["default"].nutation(jde),
      _nutation$nutation2 = _slicedToArray(_nutation$nutation, 2),
      Δψ = _nutation$nutation2[0],
      Δε = _nutation$nutation2[1];

  λ += Δψ;
  var ε = _nutation["default"].meanObliquity(jde) + Δε;
  return new _coord["default"].Ecliptic(λ, β).toEquatorial(ε); // Meeus gives a formula for elongation but doesn't spell out how to
  // obtaion term λ0 and doesn't give an example solution.
}
/**
 * Elements holds keplerian elements.
 */


var Elements = /*#__PURE__*/function () {
  /*
  Axis  float64 // Semimajor axis, a, in AU
  Ecc   float64 // Eccentricity, e
  Inc   float64 // Inclination, i, in radians
  ArgP  float64 // Argument of perihelion, ω, in radians
  Node  float64 // Longitude of ascending node, Ω, in radians
  TimeP float64 // Time of perihelion, T, as jde
  */
  function Elements(axis, ecc, inc, argP, node, timeP) {
    _classCallCheck(this, Elements);

    var o = {};

    if (_typeof(axis) === 'object') {
      o = axis;
    }

    this.axis = o.axis || axis;
    this.ecc = o.ecc || ecc;
    this.inc = o.inc || inc;
    this.argP = o.argP || argP;
    this.node = o.node || node;
    this.timeP = o.timeP || timeP;
  }
  /**
   * Position returns observed equatorial coordinates of a body with Keplerian elements.
   *
   * Argument e must be a valid V87Planet object for Earth.
   *
   * Results are right ascension and declination α and δ, and elongation ψ,
   * all in radians.
   */


  _createClass(Elements, [{
    key: "position",
    value: function position(jde, earth) {
      var _this = this;

      // (α, δ, ψ float64) {
      // (33.6) p. 227
      var n = _base["default"].K / this.axis / Math.sqrt(this.axis);
      var sε = _base["default"].SOblJ2000;
      var cε = _base["default"].COblJ2000;

      var _base$sincos9 = _base["default"].sincos(this.node),
          _base$sincos10 = _slicedToArray(_base$sincos9, 2),
          sΩ = _base$sincos10[0],
          cΩ = _base$sincos10[1];

      var _base$sincos11 = _base["default"].sincos(this.inc),
          _base$sincos12 = _slicedToArray(_base$sincos11, 2),
          si = _base$sincos12[0],
          ci = _base$sincos12[1]; // (33.7) p. 228


      var F = cΩ;
      var G = sΩ * cε;
      var H = sΩ * sε;
      var P = -sΩ * ci;
      var Q = cΩ * ci * cε - si * sε;
      var R = cΩ * ci * sε + si * cε; // (33.8) p. 229

      var A = Math.atan2(F, P);
      var B = Math.atan2(G, Q);
      var C = Math.atan2(H, R);
      var a = Math.hypot(F, P);
      var b = Math.hypot(G, Q);
      var c = Math.hypot(H, R);

      var f = function f(jde) {
        // (x, y, z float64) {
        var M = n * (jde - _this.timeP);
        var E;

        try {
          E = _kepler["default"].kepler2b(_this.ecc, M, 15);
        } catch (e) {
          E = _kepler["default"].kepler3(_this.ecc, M);
        }

        var ν = _kepler["default"].trueAnomaly(E, _this.ecc);

        var r = _kepler["default"].radius(E, _this.ecc, _this.axis); // (33.9) p. 229


        var x = r * a * Math.sin(A + _this.argP + ν);
        var y = r * b * Math.sin(B + _this.argP + ν);
        var z = r * c * Math.sin(C + _this.argP + ν);
        return {
          x: x,
          y: y,
          z: z
        };
      };

      return astrometricJ2000(f, jde, earth);
    }
  }]);

  return Elements;
}();
/**
 * AstrometricJ2000 is a utility function for computing astrometric coordinates.
 *
 * It is used internally and only exported so that it can be used from
 * multiple packages.  It is not otherwise expected to be used.
 *
 * Argument f is a function that returns J2000 equatorial rectangular
 * coodinates of a body.
 *
 * Results are J2000 right ascention, declination, and elongation.
 */


exports.Elements = Elements;

function astrometricJ2000(f, jde, earth) {
  // (f func(float64)  (x, y, z float64), jde float64, e *pp.V87Planet) (α, δ, ψ float64)
  var sol = _solarxyz["default"].positionJ2000(earth, jde);

  var _ref3 = [sol.x, sol.y, sol.z],
      X = _ref3[0],
      Y = _ref3[1],
      Z = _ref3[2];
  var ξ;
  var η;
  var ζ;
  var Δ;

  function fn() {
    var τ = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;

    // (33.10) p. 229
    var _f = f(jde - τ),
        x = _f.x,
        y = _f.y,
        z = _f.z;

    ξ = X + x;
    η = Y + y;
    ζ = Z + z;
    Δ = Math.sqrt(ξ * ξ + η * η + ζ * ζ);
  }

  fn();

  var τ = _base["default"].lightTime(Δ);

  fn(τ);
  var α = Math.atan2(η, ξ);

  if (α < 0) {
    α += 2 * Math.PI;
  }

  var δ = Math.asin(ζ / Δ);
  var R0 = Math.sqrt(X * X + Y * Y + Z * Z);
  var ψ = Math.acos((ξ * X + η * Y + ζ * Z) / R0 / Δ);
  return new _base["default"].Coord(α, δ, undefined, ψ);
}
/**
 * Velocity returns instantaneous velocity of a body in elliptical orbit around the Sun.
 *
 * Argument a is the semimajor axis of the body, r is the instaneous distance
 * to the Sun, both in AU.
 *
 * Result is in Km/sec.
 */


function velocity(a, r) {
  // (a, r float64)  float64
  return 42.1219 * Math.sqrt(1 / r - 0.5 / a);
}
/**
 * Velocity returns the velocity of a body at aphelion.
 *
 * Argument a is the semimajor axis of the body in AU, e is eccentricity.
 *
 * Result is in Km/sec.
 */


function vAphelion(a, e) {
  // (a, e float64)  float64
  return 29.7847 * Math.sqrt((1 - e) / (1 + e) / a);
}
/**
 * Velocity returns the velocity of a body at perihelion.
 *
 * Argument a is the semimajor axis of the body in AU, e is eccentricity.
 *
 * Result is in Km/sec.
 */


function vPerihelion(a, e) {
  // (a, e float64)  float64
  return 29.7847 * Math.sqrt((1 + e) / (1 - e) / a);
}
/**
 * Length1 returns Ramanujan's approximation for the length of an elliptical
 * orbit.
 *
 * Argument a is semimajor axis, e is eccentricity.
 *
 * Result is in units used for semimajor axis, typically AU.
 */


function length1(a, e) {
  // (a, e float64)  float64
  var b = a * Math.sqrt(1 - e * e);
  return Math.PI * (3 * (a + b) - Math.sqrt((a + 3 * b) * (3 * a + b)));
}
/**
 * Length2 returns an alternate approximation for the length of an elliptical
 * orbit.
 *
 * Argument a is semimajor axis, e is eccentricity.
 *
 * Result is in units used for semimajor axis, typically AU.
 */


function length2(a, e) {
  // (a, e float64)  float64
  var b = a * Math.sqrt(1 - e * e);
  var s = a + b;
  var p = a * b;
  var A = s * 0.5;
  var G = Math.sqrt(p);
  var H = 2 * p / s;
  return Math.PI * (21 * A - 2 * G - 3 * H) * 0.125;
}
/**
 * Length3 returns the length of an elliptical orbit.
 *
 * Argument a is semimajor axis, e is eccentricity.
 *
 * Result is exact, and in units used for semimajor axis, typically AU.
 */

/* As Meeus notes, Length4 converges faster.  There is no reason to use
this function
export function length3 (a, e) { // (a, e float64)  float64
  const sum0 = 1.0
  const e2 = e * e
  const term = e2 * 0.25
  const sum1 = 1.0 - term
  const nf = 1.0
  const df = 2.0
  while (sum1 !== sum0) {
    term *= nf
    nf += 2
    df += 2
    term *= nf * e2 / (df * df)
    sum0 = sum1
    sum1 -= term
  }
  return 2 * Math.PI * a * sum0
} */

/**
 * Length4 returns the length of an elliptical orbit.
 *
 * Argument a is semimajor axis, e is eccentricity.
 *
 * Result is exact, and in units used for semimajor axis, typically AU.
 */


function length4(a, e) {
  // (a, e float64)  float64
  var b = a * Math.sqrt(1 - e * e);
  var m = (a - b) / (a + b);
  var m2 = m * m;
  var sum0 = 1.0;
  var term = m2 * 0.25;
  var sum1 = 1.0 + term;
  var nf = -1.0;
  var df = 2.0;

  while (sum1 !== sum0) {
    nf += 2;
    df += 2;
    term *= nf * nf * m2 / (df * df);
    sum0 = sum1;
    sum1 += term;
  }

  return 2 * Math.PI * a * sum0 / (1 + m);
}

var _default = {
  position: position,
  Elements: Elements,
  astrometricJ2000: astrometricJ2000,
  velocity: velocity,
  vAphelion: vAphelion,
  vPerihelion: vPerihelion,
  length1: length1,
  length2: length2,
  // length3,
  length4: length4
};
exports["default"] = _default;
