// astronomia/lib/node.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ellipticAscending = ellipticAscending;
exports.ellipticDescending = ellipticDescending;
exports.el = el;
exports.parabolicAscending = parabolicAscending;
exports.parabolicDescending = parabolicDescending;
exports.pa = pa;
exports["default"] = void 0;

var _base = _interopRequireDefault(require("./base"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/**
 * EllipticAscending computes time and distance of passage through the ascending node of a body in an elliptical orbit.
 *
 * Argument axis is semimajor axis in AU, ecc is eccentricity, argP is argument
 * of perihelion in radians, timeP is time of perihelion as a jd.
 *
 * Result is jde of the event and distance from the sun in AU.
 */
function ellipticAscending(axis, ecc, argP, timeP) {
  // (axis, ecc, argP, timeP float64)  (jde, r float64)
  return el(-argP, axis, ecc, timeP);
}
/**
 * EllipticAscending computes time and distance of passage through the descending node of a body in an elliptical orbit.
 *
 * Argument axis is semimajor axis in AU, ecc is eccentricity, argP is argument
 * of perihelion in radians, timeP is time of perihelion as a jd.
 *
 * Result is jde of the event and distance from the sun in AU.
 */


function ellipticDescending(axis, ecc, argP, timeP) {
  // (axis, ecc, argP, timeP float64)  (jde, r float64)
  return el(Math.PI - argP, axis, ecc, timeP);
}

function el(ν, axis, ecc, timeP) {
  // (ν, axis, ecc, timeP float64)  (jde, r float64)
  var E = 2 * Math.atan(Math.sqrt((1 - ecc) / (1 + ecc)) * Math.tan(ν * 0.5));

  var _base$sincos = _base["default"].sincos(E),
      _base$sincos2 = _slicedToArray(_base$sincos, 2),
      sE = _base$sincos2[0],
      cE = _base$sincos2[1];

  var M = E - ecc * sE;
  var n = _base["default"].K / axis / Math.sqrt(axis);
  var jde = timeP + M / n;
  var r = axis * (1 - ecc * cE);
  return [jde, r];
}
/**
 * ParabolicAscending computes time and distance of passage through the ascending node of a body in a parabolic orbit.
 *
 * Argument q is perihelion distance in AU, argP is argument of perihelion
 * in radians, timeP is time of perihelion as a jd.
 *
 * Result is jde of the event and distance from the sun in AU.
 */


function parabolicAscending(q, argP, timeP) {
  // (q, argP, timeP float64)  (jde, r float64)
  return pa(-argP, q, timeP);
}
/**
 * ParabolicDescending computes time and distance of passage through the descending node of a body in a parabolic orbit.
 *
 * Argument q is perihelion distance in AU, argP is argument of perihelion
 * in radians, timeP is time of perihelion as a jd.
 *
 * Result is jde of the event and distance from the sun in AU.
 */


function parabolicDescending(q, argP, timeP) {
  // (q, argP, timeP float64)  (jde, r float64)
  return pa(Math.PI - argP, q, timeP);
}

function pa(ν, q, timeP) {
  // (ν, q, timeP float64)  (jde, r float64)
  var s = Math.tan(ν * 0.5);
  var jde = timeP + 27.403895 * s * (s * s + 3) * q * Math.sqrt(q);
  var r = q * (1 + s * s);
  return [jde, r];
}

var _default = {
  ellipticAscending: ellipticAscending,
  ellipticDescending: ellipticDescending,
  el: el,
  parabolicAscending: parabolicAscending,
  parabolicDescending: parabolicDescending,
  pa: pa
};
exports["default"] = _default;
