// astronomia/lib/stellar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sum = sum;
exports.sumN = sumN;
exports.ratio = ratio;
exports.difference = difference;
exports.absoluteByParallax = absoluteByParallax;
exports.absoluteByDistance = absoluteByDistance;
exports["default"] = void 0;

function _createForOfIteratorHelper(o) { if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (o = _unsupportedIterableToArray(o))) { var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var it, normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

/**
 * @copyright 2013 Sonia Keys
 * @copyright 2016 commenthol
 * @license MIT
 * @module stellar
 */

/**
 * Stellar: Chapter 56, Stellar Magnitudes.
 */

/**
 * Sum returns the combined apparent magnitude of two stars.
 */
function sum(m1, m2) {
  // (m1, m2 float64)  float64
  var x = 0.4 * (m2 - m1);
  return m2 - 2.5 * Math.log10(Math.pow(10, x) + 1);
}
/**
 * SumN returns the combined apparent magnitude of a number of stars.
 */


function sumN(m) {
  // (m ...float64)  float64
  var s = 0;

  var _iterator = _createForOfIteratorHelper(m),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var mi = _step.value;
      s += Math.pow(10, -0.4 * mi);
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  return -2.5 * Math.log10(s);
}
/**
 * Ratio returns the brightness ratio of two stars.
 *
 * Arguments m1, m2 are apparent magnitudes.
 */


function ratio(m1, m2) {
  // (m1, m2 float64)  float64
  var x = 0.4 * (m2 - m1);
  return Math.pow(10, x);
}
/**
 * Difference returns the difference in apparent magnitude of two stars
 * given their brightness ratio.
 */


function difference(ratio) {
  // (ratio float64)  float64
  return 2.5 * Math.log10(ratio);
}
/**
 * AbsoluteByParallax returns absolute magnitude given annual parallax.
 *
 * Argument m is apparent magnitude, π is annual parallax in arc seconds.
 */


function absoluteByParallax(m, π) {
  // (m, π float64)  float64
  return m + 5 + 5 * Math.log10(π);
}
/**
 * AbsoluteByDistance returns absolute magnitude given distance.
 *
 * Argument m is apparent magnitude, d is distance in parsecs.
 */


function absoluteByDistance(m, d) {
  // (m, d float64)  float64
  return m + 5 - 5 * Math.log10(d);
}

var _default = {
  sum: sum,
  sumN: sumN,
  ratio: ratio,
  difference: difference,
  absoluteByParallax: absoluteByParallax,
  absoluteByDistance: absoluteByDistance
};
exports["default"] = _default;
